### ISSUE TITLE
Dei freezes after 401

### ISSUE THREAD

**Issue by teknolojia** [flutterchina/dio#1113]
### New Issue Checklist

- [ x] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

### Issue Info

| Info                           | Value                                                 |      |
| ------------------------------ | ----------------------------------------------------- | ---- |
| Platform Name                  | e.g. flutter / ios / android                          |   flutter   |
| Platform Version               | e.g.  1.5.0 / 12.0 / 9.0                              |      |
| Dio Version                    | e.g.  2.1.0 / 1.0.17                                  |   3.0.10   |
| Android Studio / Xcode Version | e.g. Android Studio 3.3.2 / Xcode 10.2.1              |      |
| Repro rate                     | e.g. all the time (100%) / sometimes x% / only once   |   100%   |
| Repro with our demo prj        | e.g. does it happen with our demo project?            |      |
| Demo project link              | e.g. link to a demo project that highlights the issue |      |

### Issue Description and Steps

I am using below code to navigate to auth page when server responds with **401** so that user may enter credentials to get valid session again. And it does works quite  well as intended. However when user is on **AuthPage** and submits request request is fired but no response is handled. I think it has something todo with dio locks but I don't know when to lock and unlock

```dart
  dio.interceptors.add(
    InterceptorsWrapper(
      onError: (error) async {
        //If user is unauthorized
        if (error?.response?.statusCode == 401) {
          //Clearing all shared preferences
          final prefs = await SharedPreferences.getInstance();
          await prefs.clear();

          Application.navigatorKey.currentState.pushAndRemoveUntil(
            MaterialPageRoute(
              builder: (context) => AuthPage(),
            ),
            (route) => false,
          );
        }
      },
    ),
  );
```
Any comment/suggestion/help will be appreciated

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.
